libcst/codemod/commands/convert_type_comments.py

```diff
@@ -0,0 +1,163 @@
+from typing import Optional, Pattern, Set, Union
+
+import libcst as cst
+from libcst.codemod import CodemodContext, VisitorBasedCodemodCommand
+
+
+@functools.lru_cache()
+def _empty_module() -> cst.Module:
+    return cst.parse_module("")
+
+
+def _code_for_node(node: cst.CSTNode) -> str:
+    return _empty_module().code_for_node(node)
+
+
+def _ast_for_node(node: cst.CSTNode) -> ast.Module:
+    code = _code_for_node(node)
+    return ast.parse(code, type_comments=True)
+
+
+def _assign_type_comment(node: cst.SimpleStatementLine) -> Optional[str]:
+    return _ast_for_node(node).body[0].type_comment
+
+
+@functools.lru_cache()
+def _builtins() -> Set[str]:
+    return set(dir(builtins))
+
+
+def _is_builtin(annotation: str) -> bool:
+    return annotation in _builtins()
+
+
+def _convert_annotation(raw: str) -> cst.Annotation:
+    # Convert annotation comments to string annotations to be safe,
+    # otherwise runtime errors would be common.
+    #
+    # Special-case builtins to reduce the amount of quoting noise.
+    #
+    # NOTE: we could potentially detect more cases for skipping quotes
+    # using ScopeProvider, which would make the output prettier.
+    if _is_builtin(raw):
+        return cst.Annotation(annotation=cst.Name(value=raw))
+    else:
+        return cst.Annotation(annotation=cst.SimpleString(f'"{raw}"'))
+
+
+class ConvertTypeComments(VisitorBasedCodemodCommand):
+    """
+    Codemod that converts type comments, as described in
+    https://www.python.org/dev/peps/pep-0484/#type-comments,
+    into PEP 526 annotated assignments.
+    """
+
+    TYPE_COMMENT_REGEX: Pattern[str] = re.compile("# *type: *[^ ]+")
+    TYPE_IGNORE_PREFIX: str = "# type: ignore"
+
+    _should_strip_type_comments: bool = False
+
+    def __init__(self, context: CodemodContext) -> None:
+        if (sys.version_info.major, sys.version_info.minor) < (3, 8):
+            # The ast module did not get `type_comments` until Python 3.7.
+            # In 3.6, we should error than silently running a nonsense codemod.
+            #
+            # NOTE: it is possible to use the typed_ast library for 3.6, but
+            # this is not a high priority right now. See, e.g., the
+            # mypy.fastparse module.
+            raise NotImplementedError(
+                "You are trying to run ConvertTypeComments on a "
+                + "python version without type comment support. Please "
+                + "try using python 3.8+ to run your codemod."
+            )
+        super().__init__(context)
+
+    def _is_type_comment(self, comment: Optional[cst.Comment]) -> bool:
+        return (
+            comment is not None
+            and self.TYPE_COMMENT_REGEX.match(comment.value) is not None
+            and not comment.value.startswith(self.TYPE_IGNORE_PREFIX)
+        )
+
+    def _should_strip_comment(self, comment: Optional[cst.Comment]) -> bool:
+        return self._should_strip_type_comments and self._is_type_comment(comment)
```

Comment:
This seems to be an unreliable way of detecting whether a comment is a type comment or not. Wouldn't it be possible to detect whether to strip comments based on whether the containing ast node has `type_comments` attribute set instead?

Reply:
Hmm... I don't think it's going to be possible to do this in general using a visitor pattern. I could try extracting type information in a `visit_` method on the parent, but that won't work later on when I handle more cases because the same problem is going to come up with function defs, with statements, and for statements.

I guess the answer is going to be to transform whitespace manually, which I was hoping to avoid because in the FunctionDef case it's not always in the same place. But that will probably be much more reliable even though it's ugly.